app/models/piece.rb

```diff
@@ -1,8 +1,147 @@
-class Piece < ApplicationRecord
-  belongs_to :user
-  belongs_to :game
-
-  def self.types
-    %w[King Queen Rook Bishop Knight Pawn]
-  end
-end
+class Piece < ApplicationRecord
+  belongs_to :user
+  belongs_to :game
+
+  def self.types
+    %w[King Queen Rook Bishop Knight Pawn]
+  end
+
+  def obstructed?(destination_x, destination_y)
+    raise 'Error: Invalid Input' if destination_y > 7 || destination_x > 7 || destination_y < 0 || destination_x < 0
+    invalid_move_tests =
+      [
+        horizontal_move?(destination_y),
+        vertical_move?(destination_x),
+        diagonal_up_and_right_move?(destination_x, destination_y),
+        diagonal_down_and_left_move?(destination_x, destination_y),
+        diagonal_up_and_left_move?(destination_x, destination_y),
+        diagonal_down_and_right_move?(destination_x, destination_y)
+      ]
+    raise 'Error: Invalid Input' unless invalid_move_tests.include?(true)
+    destination_tests = []
+    invalid_move_tests.each_with_index do |test, index|
+      if test == true && index == 0
+        destination_tests << horizontal_obstruction_left?(destination_x)
+        destination_tests << horizontal_obstruction_right?(destination_x)
+      elsif test == true && index == 1
+        destination_tests << vertical_obstruction_up?(destination_y)
+        destination_tests << vertical_obstruction_down?(destination_y)
+      elsif test == true && index == 2
+        destination_tests << diagonal_obstruction_up_right?(destination_y)
+      elsif test == true && index == 3
+        destination_tests << diagonal_obstruction_down_left?(destination_y)
+      elsif test == true && index == 4
+        destination_tests << diagonal_obstruction_up_left?(destination_y)
+      elsif test == true && index == 5
+        destination_tests << diagonal_obstruction_down_right?(destination_y)
+      end
+    end
+    return true if destination_tests.include?(true)
+    false
+  end
+
+  def horizontal_move?(destination_y)

```

Comment:
All of these methods can be `private`

(Add `private` above this method and all methods below it will then be private)

http://culttt.com/2015/06/03/the-difference-between-public-protected-and-private-methods-in-ruby/